presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/MetadataResolver.java

```diff
@@ -87,4 +90,24 @@ default boolean isMaterializedView(QualifiedObjectName viewName)
     {
         return getMaterializedView(viewName).isPresent();
     }
+
+    /**
+     * Returns list of registered types.
+     */
+    List<Type> getTypes();
+
+    /**
+     * Returns list of registered parametric types.
+     */
+    List<ParametricType> getParametricTypes();
+
+    /**
+     * Returns list of registered functions.
+     */
+    List<SqlFunction> listFunctions();
```

Comment:
Depends on how we design our analyzer in the long term, having a list of function could be useful or might not be useful. Indeed, we will have infinite number of functions because types can be parametric (e.g., array of array of array of .... of array of int) leading to infinite number of functions. So checking the existence and return type of a function seems more important than having a list of functions.

Reply:
This is really an open question at this point. I have tried to mimic current crux behavior. It relies on show functions to get list of all functions. We had few options to choose from while designing this, eg versioning support. We believe checking for each entry might actually be slower. We probably would need to revisit this when we integrate with bindingcontext.